https://github.com/wahani/modules/issues/13

wahani opened the pull request:
Hope this is somewhat close to what you want!

If this doesn't do what you want, then I would need to see how the 'things' in the folder 'modules' are called. So how would you integrate these shiny-modules exactly. Then I can debug.

wahani committed:
https://github.com/wahani/modules/issues/13

BetweenTwoTests commented:
Thanks @wahani !

I am not sure what you mean by "see how the 'things' in the folder 'modules' are called". If you mean how do I know if Shiny module functions are are called lagUI and lagSERVER, then this can be done via separate file that keeps track of functions, by convention, or by reading off the names in module object once read into R with module::use()

I am always wary of overwriting the named functions of other packages (i.e. `library` and `use` in this case). I will test if this is what we want!

One question is, if we have a `modules/lag/lag.R` script that uses `::` without directly loading the package via `library(..)`, then does this solution deal with that too? It doesn't seem like it?

wahani commented:
I think we are not there yet. I forgot about something...

It would be nice to have an example how you use `lagSERVER` and `lagUI` so I can see it in action.

wahani commented:
We are not overriding `library`, only masking it, although this may be what you mean. I would not recommend doing it either. However, if your authors are using `library` (and that causes problems), you can (1) eliminate all problems by using R-packages or (2) by telling them to use `::` instead and never `library`, or (3) tell them to use `modules::import`, or (4) don't tell them anything and mask `base::library`.

You do not need to call `library` if your users all faithfully use `::`.

BetweenTwoTests commented:
I planned to just keep it in a R object and extract raw string values. then access the module object

```
m <- use(...)
list_of_modules <- c("lagUI", "lagSERVER")
# calling lagUI server function via module
m[[ list_of_modules[1] ]](...)

# calling lagUI server function via module
m[[ list_of_modules[2] ]](...)
```

Something along this manner. I'll update with example.

-------

Unfortunately, the plan is to work with beginner R users. Having them to learn library(modules) is a bit out of reach for many :) I guess overriding is the best solution given the problem I am trying to solve.

wahani commented:
>It would be nice to have modules::use() to reconcile "R scripts built via the normal R way" and translate it into library(modules). However, I see that this may be impossible / bad idea. Should I manually edit modules/lag/lag.R to be compliant library(module) manually? Would this be the only / best way? The goal is to make the integration of R script made by other users to seamlessly integrate into the Shiny app.

Maybe you can have a 'preprocessing' of the authors code. So that your `components/03_tab_3.R` is automatically generated by a script (as part of your global.R or on top of app.R). There you can then add the line `library <- modules::import` as first line in the module. Or you replace all `library` with `modules::import` (this may be more difficult). If the scope of this exercise is limited, I think it is safer than it sounds ;)

wahani commented:
Something along the lines of 

```r
preprocess <- function(fileName) {
  txt <- readLines(fileName)
  txt <- c("library <- modules::import", txt)
  writeLines(txt, paste0("components/", basename(fileName)))
}

lapply(list.files("modules", full.names = TRUE, include.dirs = FALSE), preprocess)
```

Hope you get the idea. This would be the preprocessing.

wahani commented:
With that approach, I would then recommend using the `use` function from https://github.com/wahani/modules/issues/13#issuecomment-545079669 so you do not have to worry about the stats and friends packages.

BetweenTwoTests committed:
- Rename modules/ R scripts to be more descriptive and indicative of whether library(modules) is used
- Add example of Shiny module with both dplyr and stats
- Add text about what the desired output is
- remove lagUI and lagSERVER on stop

BetweenTwoTests committed:
- Add use() variant as Leon wanted
- moved *_use() functions to global.R
- Added more examples to demonstrate desired output
- Edit app to show more demo functionality

BetweenTwoTests committed:
mimicked module loading should use non-library(modules) version of R script, not library(modules) scripts

BetweenTwoTests committed:
use wahani's suggestions

wahani committed:
Merge pull request #1 from BetweenTwoTests/wahani-patch-1

Extending modules for using Shiny modules
